processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerColumnar.java

```diff
@@ -488,13 +488,19 @@ private NodeHolder processDataRows(List<CarbonRow> dataRows)
   public void finish() throws CarbonDataWriterException {
     // still some data is present in stores if entryCount is more
     // than 0
-    producerExecutorServiceTaskList.add(producerExecutorService
-        .submit(new Producer(blockletDataHolder, dataRows, ++writerTaskSequenceCounter, true)));
-    blockletProcessingCount.incrementAndGet();
-    processedDataCount += entryCount;
-    closeWriterExecutionService(producerExecutorService);
-    processWriteTaskSubmitList(producerExecutorServiceTaskList);
-    processingComplete = true;
+    try {
+      semaphore.acquire();
```

Comment:
@watermen .....here a check is required for entryCount > 0.....because we need to acquire a semaphore lock only if total number of rows in raw data are not exactly divisible by page size...in this case only we will have some extra rows to be process by finish method else addDataToStore method will handle all the rows....Please refer the below code snippet....

public void finish() throws CarbonDataWriterException {
    // still some data is present in stores if entryCount is more
    // than 0
    if (this.entryCount > 0) {
      try {
        semaphore.acquire();
        producerExecutorServiceTaskList.add(producerExecutorService
            .submit(new Producer(blockletDataHolder, dataRows, ++writerTaskSequenceCounter, true)));
        blockletProcessingCount.incrementAndGet();
        processedDataCount += entryCount;
      } catch (InterruptedException e) {
        LOGGER.error(e, e.getMessage());
        throw new CarbonDataWriterException(e.getMessage(), e);
      }
    }
    closeWriterExecutionService(producerExecutorService);
    processWriteTaskSubmitList(producerExecutorServiceTaskList);
    processingComplete = true;